### ISSUE TITLE
Bad diagnostics on struct literal in for-loop iterator expression

### ISSUE THREAD

**Issue by abonander** [intellij-rust/intellij-rust#3039]
## Environment

* **Intellij-Rust plugin version:** 0.2.0.2471-182-nightly
* **Rust toolchain version:** rustc 1.32.0-nightly (8b096314a 2018-11-02)
* **IDE name and version (and Windows version, helpfully provided by the "About" dialog):**
IntelliJ IDEA 2018.2.2 (Ultimate Edition)
Build #IU-182.4129.33, built on August 21, 2018
Licensed to Austin Bonander
Subscription is active until May 11, 2019
JRE: 1.8.0_152-release-1248-b8 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Windows 10 10.0

## Problem description

Rust as of 1.32 Nightly (and probably for a long time before) requires parenthesis around struct
literals when used in the iterator expression of a `for` loop:

```rust
struct SomeIter {
    foo: ()
}

impl Iterator for SomeIter { /* irrelevant */ }

fn main() {
    // error: expected expression, found struct
    for val in SomeIter { foo: () } {}
}
```

IntelliJ-Rust fails to resolve the reference to `SomeIter` in this form:
![image](https://user-images.githubusercontent.com/3198595/48467738-a0988800-e79e-11e8-9570-cb95ade4d4c7.png)

The compiler helpfully suggests wrapping it in parenthesis, but then IntelliJ-Rust displays a false lint warning about them:
![image](https://user-images.githubusercontent.com/3198595/48467811-e05f6f80-e79e-11e8-9f6d-e1d721f4e6eb.png)
(this form compiles successfully without lint warnings from the compiler)

**Comment by abonander** [comment]
@vlad20012 the error in the first screenshot could be improved as well; it doesn't make sense that it can't resolve the `SomeIter` reference since it's in-scope.

**Comment by vlad20012** [comment]
@abonander It's too complex task for now, because our resolve engine literally can't resolve it (because of different types/values namespaces) and should be improved first =)

**Comment by abonander** [comment]
@vlad20012 I mean it should show a syntax error (struct initializers need to be wrapped in parenthesis) instead of the resolve error.

**Comment by vlad20012** [comment]
I don't see **syntax** errors resorted by rustc. Your example can be parsed successfully
```rust
struct SomeIter {
    foo: ()
}

fn main() {
    // error: expected expression, found struct
    for val in SomeIter { foo: () } {}
}
```

```
error[E0423]: expected value, found struct `SomeIter`
 --> src/main.rs:7:16
  |
7 |     for val in SomeIter { foo: () } {}
  |                ^^^^^^^^ did you mean `(SomeIter { /* fields */ })`?

error[E0425]: cannot find value `foo` in this scope
 --> src/main.rs:7:27
  |
7 |     for val in SomeIter { foo: () } {}
  |                           ^^^
  |                           |
  |                           not found in this scope
  |                           expecting a type here because of type ascription

error[E0658]: type ascription is experimental (see issue #23416)
 --> src/main.rs:7:27
  |
7 |     for val in SomeIter { foo: () } {}
  |                           ^^^^^^^
  |
  = help: add #![feature(type_ascription)] to the crate attributes to enable
```

**Comment by vlad20012** [comment]
I think we should report E0423 just like rustc and suggest to wrap it with parenthesis, just like rustc suggests. But it is rather complex problem, I described it in the previous comment :)

**Comment by abonander** [comment]
@vlad20012 so reopen this issue or a new one?

**Comment by vlad20012** [comment]
I think new one. You can attach my example as well.
![image](https://user-images.githubusercontent.com/3221931/48822024-8f3a1780-ed6c-11e8-9f67-ebbb496fa98e.png)
